libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,470 @@
+         });
+      });
+
+      add_contract_tables_to_snapshot(snapshot);
+
+      authorization.add_to_snapshot(snapshot);
+      resource_limits.add_to_snapshot(snapshot);
+   }
+
+   void combined_database::read_from_snapshot(const snapshot_reader_ptr& snapshot, uint32_t blog_start,
+                                              uint32_t blog_end, backing_store_type backing_store_,
+                                              eosio::chain::authorization_manager&                    authorization,
+                                              eosio::chain::resource_limits::resource_limits_manager& resource_limits,
+                                              eosio::chain::fork_database& fork_db, eosio::chain::block_state_ptr& head,
+                                              uint32_t&                          snapshot_head_block,
+                                              const eosio::chain::chain_id_type& chain_id) {
+      chain_snapshot_header header;
+      snapshot->read_section<chain_snapshot_header>([this, &header](auto& section) {
+         section.read_row(header, db);
+         header.validate();
+      });
+
+      db.create<kv_db_config_object>([](auto&) {});
+      set_backing_store(backing_store_);
+
+      { /// load and upgrade the block header state
+         block_header_state head_header_state;
+         using v2 = legacy::snapshot_block_header_state_v2;
+
+         if (std::clamp(header.version, v2::minimum_version, v2::maximum_version) == header.version) {
+            snapshot->read_section<block_state>([this, &head_header_state](auto& section) {
+               legacy::snapshot_block_header_state_v2 legacy_header_state;
+               section.read_row(legacy_header_state, db);
+               head_header_state = block_header_state(std::move(legacy_header_state));
+            });
+         } else {
+            snapshot->read_section<block_state>(
+                  [this, &head_header_state](auto& section) { section.read_row(head_header_state, db); });
+         }
+
+         snapshot_head_block = head_header_state.block_num;
+         EOS_ASSERT(blog_start <= (snapshot_head_block + 1) && snapshot_head_block <= blog_end, block_log_exception,
+                    "Block log is provided with snapshot but does not contain the head block from the snapshot nor a "
+                    "block right after it",
+                    ("snapshot_head_block", snapshot_head_block)("block_log_first_num",
+                                                                 blog_start)("block_log_last_num", blog_end));
+
+         fork_db.reset(head_header_state);
+         head                = fork_db.head();
+         snapshot_head_block = head->block_num;
+      }
+
+      controller_index_set::walk_indices([this, &snapshot, &header](auto utils) {
+         using value_t = typename decltype(utils)::index_t::value_type;
```

Comment:
same as above.

Reply:
Yes, there are four such constructs in this file. I assume they're intentional.